Protobuf 3.5.x Fails to Compile on GCC 7+ With -std Less Than C++11

**Version:** v3.5.1
**Language:** C++
**Operating System:** Linux - Debian Testing
**Compiler:** GCC 7.3.0 (Debian GCC 7.3.0-5, specifically)

**Steps to reproduce the behavior:**
1. Compile Protobuf with the -std=gnu++98 flag set

**Expected result:**
Protobuf compiles, as C++98 is still supported on the 3.5.x releases.

**Actual result:**
Protobuf fails to compile with the following error:

```
In file included from ./src/google/protobuf/stubs/common.h:45:0,
                 from ./src/google/protobuf/map.h:45,
                 from ./src/google/protobuf/generated_message_table_driven.h:34,
                 from ./src/google/protobuf/generated_message_table_driven_lite.h:34,
                 from ./src/google/protobuf/generated_message_table_driven_lite.cc:31:
./src/google/protobuf/map_entry_lite.h: In member function ‘bool google::protobuf::internal::MapEntryImpl<Derived, Base, Key, Value, kKeyFieldType, kValueFieldType, default_enum_value>::MergePartialFromCodedStream(google::protobuf::io::CodedInputStream*)’:
./src/google/protobuf/stubs/port.h:247:57: warning: c++11 attributes only available with -std=c++11 or -std=gnu++11
 # define GOOGLE_FALLTHROUGH_INTENDED [[gnu::fallthrough]]
                                                         ^
./src/google/protobuf/map_entry_lite.h:204:11: note: in expansion of macro ‘GOOGLE_FALLTHROUGH_INTENDED’
           GOOGLE_FALLTHROUGH_INTENDED;
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Investigation:**
The issue is that port.h assumes that `[[gnu::fallthrough]]` is always available on GCC 7+. While the `gnu::fallthrough` annotation is available from 7 on, the C++ annotations feature is *not* available unless the standard in use is C++11 or newer.

This problem exists only on Protobuf 3.5.x releases. Earlier releases did not have this code path. Later releases require C++11 or greater and have removed `GOOGLE_FALLTHROUGH_INTENDED`.

--- a/src/google/protobuf/stubs/port.h
+++ b/src/google/protobuf/stubs/port.h
@@ -243,7 +243,7 @@ static const uint64 kuint64max = GOOGLE_ULONGLONG(0xFFFFFFFFFFFFFFFF);
      __has_cpp_attribute(clang::fallthrough)
 #  define GOOGLE_FALLTHROUGH_INTENDED [[clang::fallthrough]]
 # endif
-#elif defined(__GNUC__) && __GNUC__ > 6
+#elif defined(__GNUC__) && __GNUC__ > 6 && __cpp_attributes >= 200809
 # define GOOGLE_FALLTHROUGH_INTENDED [[gnu::fallthrough]]
 #endif
 

